rust/otap-dataflow/crates/otap/src/otap_grpc.rs

```diff
@@ -224,162 +168,186 @@ impl ArrowTracesService for ArrowTracesServiceImpl {
-                    log::error!("Error sending BatchStatus response: {e:?}");
-                })?;
-
-                return Ok(());
+        }
+
+        async fn process_batch(&mut self, mut batch: BatchArrowRecords) -> BatchProcessingResult {
+            let batch_id = batch.batch_id;
+
+            let batch = match self.consumer.consume_bar(&mut batch) {
+                Ok(batch) => batch,
+                Err(e) => {
+                    log::error!("Error decoding OTAP Batch: {e:?}. Closing stream");
+                    return BatchProcessingResult::Terminate;
+                }
+            };
+
+            let batch = from_record_messages::<T>(batch);
+            let otap_batch_as_otap_arrow_records = (self.otap_batch)(batch);
+            let mut otap_pdata =
+                OtapPdata::new(Context::default(), otap_batch_as_otap_arrow_records.into());
+
+            let cancel_rx = if let Some(state) = self.state.clone() {
+                let allocation_result = state.0.lock().allocate(|| oneshot::channel());
+                let (key, rx) = match allocation_result {
+                    None => {
+                        log::error!("Too many concurrent requests");
+                        return BatchProcessingResult::Emit(BatchStatus {
+                            batch_id,
+                            status_code: StatusCode::Unavailable as i32,
+                            status_message: format!(
+                                "Pipeline processing failed: {}",
+                                "Too many concurrent requests"
+                            ),
+                        });
+                    }
+                    Some(pair) => pair,
+                };
+
+                self.effect_handler.subscribe_to(
+                    Interests::ACKS | Interests::NACKS,
+                    key.into(),
+                    &mut otap_pdata,
+                );
+                Some((SlotGuard { key, state }, rx))
+            } else {
+                None
+            };
+
+            if let Err(e) = self.effect_handler.send_message(otap_pdata).await {
+                log::error!("Failed to send to pipeline: {e}");
+                return BatchProcessingResult::Terminate;
+            };
+
+            if let Some((_cancel_guard, rx)) = cancel_rx {
+                match rx.await {
+                    Ok(Ok(())) => {}
+                    Ok(Err(nack)) => {
+                        return BatchProcessingResult::Emit(BatchStatus {
+                            batch_id,
+                            status_code: StatusCode::Unavailable as i32,
+                            status_message: format!("Pipeline processing failed: {}", nack.reason),
+                        });
+                    }
+                    Err(_) => {
+                        log::error!("Response channel closed unexpectedly");
+                        return BatchProcessingResult::Terminate;
+                    }
+                }
```

Comment:
cc @utpilla in #1324 described a feature request to implement asynchronous requests, which the `otelarrowreceiver` does from Phase 1.

Do you object to concurrent execution, meaning to allow the stream to continue reading with some number of outstanding responses pending?

Reply:
No objection.